lib/jxl/modular/encoding/encoding.cc

```diff
@@ -196,6 +196,61 @@ Status DecodeModularChannelMAANS(BitReader *br, ANSSymbolReader *reader,
           }
         }
       }
+    } else if (predictor == Predictor::Gradient && offset == 0 &&
+               multiplier == 1 && reader->RleOnly()) {
+      JXL_DEBUG_V(8, "Gradient RLE (fjxl) very fast track.");
+      const intptr_t onerow = channel.plane.PixelsPerRow();
+      uint32_t run = 0;
+      uint32_t v = 0;
+      pixel_type sv = 0;
+      for (size_t y = 0; y < channel.h; y++) {
+        pixel_type *JXL_RESTRICT r = channel.Row(y);
+
+#if true
+        const pixel_type *JXL_RESTRICT rtop = (y ? channel.Row(y - 1) : r - 1);
+        const pixel_type *JXL_RESTRICT rtopleft =
+            (y ? channel.Row(y - 1) - 1 : r - 1);
+        {
+          size_t x = 0;
+          pixel_type left = (x ? r[x - 1] : y ? *(r + x - onerow) : 0);
+          pixel_type top = (y ? *(r + x - onerow) : left);
+          pixel_type topleft = (x && y ? *(r + x - 1 - onerow) : left);
+          pixel_type guess = ClampedGradient(top, left, topleft);
+          if (run == 0) {
+            reader->ReadHybridUintClusteredRleOnly(ctx_id, br, &v, &run);
+            sv = UnpackSigned(v);
+          } else
+            run--;
+          r[x] = sv + guess;
+        }
+        for (size_t x = 1; x < channel.w; x++) {
+          pixel_type left = r[x - 1];
+          pixel_type top = rtop[x];
+          pixel_type topleft = rtopleft[x];
+          pixel_type guess = ClampedGradient(top, left, topleft);
+          if (!run) {
+            reader->ReadHybridUintClusteredRleOnly(ctx_id, br, &v, &run);
+            sv = UnpackSigned(v);
+          } else
+            run--;
+          r[x] = sv + guess;
```

Comment:
I think this may risk overflowing, which (a) UBSAN would complain about (b) is UB...

Reply:
I think only (a) is a problem. IIRC, spec says that modular values not fitting in int32_t is not a valid bitstream...

Reply:
Changed it to use pixel_type_w

Reply:
Is it significantly slower? Can we just use make_pixel?

Reply:
It's a bit faster, actually. Make_pixel is a bit slower because it does more UnpackSigned in case of runs...

Reply:
Ah but with sv a pixel_type_w this works out without overflows. NVM then, great ;)